lib/livebook/runtime/k8s.ex

```diff
@@ -0,0 +1,566 @@
+
+      _ ->
+        Logger.debug("[K8s runtime] #{name} finished in #{milliseconds}ms")
+    end
+
+    result
+  end
+
+  defp with_pod_events(caller, name, req, namespace, pod_name, fun) do
+    with_log(caller, name, fn ->
+      runtime_pid = self()
+
+      event_watcher_pid =
+        spawn(fn ->
+          event_req =
+            Req.merge(req,
+              resource_path: "api/v1/namespaces/:namespace/events/:name",
+              resource_list_path: "api/v1/namespaces/:namespace/events"
+            )
+
+          {:ok, stream} =
+            Kubereq.watch(event_req, namespace,
+              field_selectors: [
+                {"involvedObject.kind", "Pod"},
+                {"involvedObject.name", pod_name}
+              ]
+            )
+
+          stream
+          |> Enum.each(fn event ->
+            Logger.debug("[K8s runtime] Pod event: \"#{event["object"]["message"]}\"")
+            send(caller, {:runtime_connect_info, runtime_pid, event["object"]["message"]})
+          end)
+        end)
+
+      result = fun.()
+      Process.exit(event_watcher_pid, :normal)
+      result
+    end)
+  end
+
+  defp create_pod(req, config, runtime_data, remote_port) do
+    %{
+      pod_template: pod_template,
+      docker_tag: docker_tag,
+      home_pvc: home_pvc,
+      namespace: namespace
+    } = config
+
+    manifest =
+      pod_template.template
+      |> Pod.pod_from_template()
+      |> Pod.add_env_vars([
+        %{"name" => "LIVEBOOK_RUNTIME", "value" => runtime_data},
+        %{"name" => "ERL_AFLAGS", "value" => "-proto_dist inet6_tcp"},
+        %{
+          "name" => "POD_IP",
+          "valueFrom" => %{"fieldRef" => %{"apiVersion" => "v1", "fieldPath" => "status.podIP"}}
+        },
+        %{
+          "name" => "POD_NAMESPACE",
+          "valueFrom" => %{
+            "fieldRef" => %{"apiVersion" => "v1", "fieldPath" => "metadata.namespace"}
+          }
+        },
+        %{
+          "name" => "POD_NAME",
+          "valueFrom" => %{"fieldRef" => %{"apiVersion" => "v1", "fieldPath" => "metadata.name"}}
+        }
+      ])
+      |> Pod.set_docker_tag(docker_tag)
+      |> Pod.set_home_pvc(home_pvc)
+      |> Pod.set_namespace(namespace)
+      |> Pod.add_container_port(remote_port)
+
+    case Kubereq.create(req, manifest) do
+      {:ok, %{status: 201, body: %{"metadata" => %{"name" => pod_name}}}} ->
+        {:ok, pod_name}
+
+      {:ok, %{body: body}} ->
+        {:error, "could not create Pod, reason: #{body["message"]}"}
+
+      {:error, error} ->
+        {:error, "could not create Pod, reason: #{Exception.message(error)}"}
+    end
+  end
+
+  defp get_cluster_data(nil), do: %{node_base: "k8s_runtime", remote_port: 44444}
+
+  defp get_cluster_data(_kubeconfig) do
+    local_port = get_free_port!()
+    %{node_base: "remote_runtime_#{local_port}", remote_port: 44444, local_port: local_port}
+  end
+
+  defp k8s_forward_port(nil, _, _, _, _), do: :ok
```

Comment:
> This is when running Livebook itself on K8s, right? 

Yes exactly. kubeconfig is `nil` if we're running in a Kubernetes cluster:

```ex
    kubeconfig =
      if System.get_env("KUBERNETES_SERVICE_HOST"),
        do: nil,
        else: System.get_env("KUBECONFIG") || Path.join(System.user_home!(), ".kube/config")

    cluster_data = get_cluster_data(kubeconfig)
```

But also `get_cluster_data/1` is different. When running in a cluster, we set `node_base` to `"k8s_runtime"` instead of `"remote_runtime_#{local_port}"`. Later in the continue we then set `child_node = :"#{cluster_data.node_base}@#{pod_ip}"`. 
So if we're running inside a cluster, we're correctly targeting the pod's IP address while when running locally, that part doesn't matter and loopback is targeted.